### ISSUE TITLE
Deploy pre-built webstack containers to cluster

### ISSUE THREAD

**Issue by cmische** [Misch-/Kubernetes-Capstone#3]
We mentioned in our meeting that we would try and deploy a set of services to the cluster for our next meeting. Any kind of web front end that hooks up to a database and does some stuff sounds good to me.

**Comment by mjt112** [comment]
Did this example on my own machine as well as the IBM VM.
http://kubernetes.io/v1.1/examples/guestbook/README.html

1)  First step was setting up a redis master for this container in the Workbook example

```
capstone@kube01:~/kubernetes-1.1.7/examples/guestbook$ sudo kubectl create -f ./redis-master-controller.yaml
replicationcontrollers/redis-master
capstone@kube01:~/kubernetes-1.1.7/examples/guestbook$ kubectl get rc
-bash: /usr/bin/kubectl: Permission denied
capstone@kube01:~/kubernetes-1.1.7/examples/guestbook$ sudo kubectl get rc
CONTROLLER     CONTAINER(S)   IMAGE(S)   SELECTOR            REPLICAS
redis-master   master         redis      name=redis-master   1
capstone@kube01:~/kubernetes-1.1.7/examples/guestbook$ sudo kubectl get pods
NAME                 READY     STATUS    RESTARTS   AGE
redis-master-k8by5   1/1       Running   0          51s
```

This is where I realized I did not create in the proper namespace for the already made cluster but I figured this example didn't require it to be on our main namespace

```
capstone@kube01:~/kubernetes-1.1.7/examples/guestbook$ sudo kubectl get pods -o wide --all-namespaces=true
NAMESPACE     NAME                 READY     STATUS    RESTARTS   AGE       NODE
default       redis-master-k8by5   1/1       Running   0          2m        169.44.62.99
kube-system   kube-dns-v9-ul4fb    3/4       Running   4          1d        169.55.21.75
kube-system   kube-ui-v4-sxr6g     1/1       Running   0          1d        169.44.62.101
capstone@kube01:~/kubernetes-1.1.7/examples/guestbook$
```

Additional info of pod

```
capstone@kube01:~/kubernetes-1.1.7/examples/guestbook$ sudo kubectl describe pods/redis-master-k8by5
Name:               redis-master-k8by5
Namespace:          default
Image(s):           redis
Node:               169.44.62.99/169.44.62.99
Labels:             name=redis-master
Status:             Running
Reason:             
Message:            
IP:             172.16.95.2
Replication Controllers:    redis-master (1/1 replicas created)
Containers:
  master:
    Image:      redis
    State:      Running
      Started:      Wed, 10 Feb 2016 07:25:10 -0600
    Ready:      True
    Restart Count:  0
Conditions:
  Type      Status
  Ready     True
Events:
  FirstSeen             LastSeen            Count   From            SubobjectPath               Reason      Message
  Wed, 10 Feb 2016 07:24:54 -0600   Wed, 10 Feb 2016 07:24:54 -0600 1   {scheduler }        scheduled   Successfully assigned redis-master-k8by5 to 169.44.62.99
  Wed, 10 Feb 2016 07:24:58 -0600   Wed, 10 Feb 2016 07:24:58 -0600 1   {kubelet 169.44.62.99}  implicitly required container POD   pulled      Successfully pulled Pod container image "gcr.io/google_containers/pause:0.8.0"
  Wed, 10 Feb 2016 07:24:58 -0600   Wed, 10 Feb 2016 07:24:58 -0600 1   {kubelet 169.44.62.99}  implicitly required container POD   created     Created with docker id 8bf1a0d038fb
  Wed, 10 Feb 2016 07:24:58 -0600   Wed, 10 Feb 2016 07:24:58 -0600 1   {kubelet 169.44.62.99}  implicitly required container POD   started     Started with docker id 8bf1a0d038fb
  Wed, 10 Feb 2016 07:25:10 -0600   Wed, 10 Feb 2016 07:25:10 -0600 1   {kubelet 169.44.62.99}  spec.containers{master}         pulled      Successfully pulled image "redis"
  Wed, 10 Feb 2016 07:25:10 -0600   Wed, 10 Feb 2016 07:25:10 -0600 1   {kubelet 169.44.62.99}  spec.containers{master}         created     Created with docker id ad59993ec104
  Wed, 10 Feb 2016 07:25:10 -0600   Wed, 10 Feb 2016 07:25:10 -0600 1   {kubelet 169.44.62.99}  spec.containers{master}         started     Started with docker id ad59993ec104
```

2)  This is to set up the redis master service to route traffic to a single master

```
capstone@kube01:~/kubernetes-1.1.7/examples/guestbook$ sudo kubectl create -f ./redis-master-service.yaml
services/redis-master
capstone@kube01:~/kubernetes-1.1.7/examples/guestbook$ sudo kubectl get services
NAME           LABELS                                    SELECTOR            IP(S)          PORT(S)
kubernetes     component=apiserver,provider=kubernetes   <none>              192.168.3.1    443/TCP
redis-master   name=redis-master                         name=redis-master   192.168.3.12   6379/TCP
```

3)  Creating replicated pods of master with a defined # of 2 for minimum pods

```
capstone@kube01:~/kubernetes-1.1.7/examples/guestbook$ sudo kubectl create -f ./redis-slave-controller.yaml
replicationcontrollers/redis-slave
capstone@kube01:~/kubernetes-1.1.7/examples/guestbook$ sudo kubectl get pods
NAME                 READY     STATUS    RESTARTS   AGE
redis-master-k8by5   1/1       Running   0          22m
redis-slave-806w4    1/1       Running   0          19s
redis-slave-bdd04    1/1       Running   0          19s
capstone@kube01:~/kubernetes-1.1.7/examples/guestbook$
```

4)  Creating service to proxy connections to the slaves

```
capstone@kube01:~/kubernetes-1.1.7/examples/guestbook$ sudo kubectl create -f ./redis-slave-service.yaml
services/redis-slave
capstone@kube01:~/kubernetes-1.1.7/examples/guestbook$ sudo kubectl get services
NAME           LABELS                                    SELECTOR            IP(S)          PORT(S)
kubernetes     component=apiserver,provider=kubernetes   <none>              192.168.3.1    443/TCP
redis-master   name=redis-master                         name=redis-master   192.168.3.12   6379/TCP
redis-slave    name=redis-slave                          name=redis-slave    192.168.3.98   6379/TCP
```

5)  Creating the frontend pod which is a php server to talk to the slave or master services

```
capstone@kube01:~/kubernetes-1.1.7/examples/guestbook$ sudo vim frontend-controller.yaml
capstone@kube01:~/kubernetes-1.1.7/examples/guestbook$ sudo kubectl create -f ./frontend-controller.yaml
replicationcontrollers/frontend
capstone@kube01:~/kubernetes-1.1.7/examples/guestbook$ sudo kubectl get rc
CONTROLLER     CONTAINER(S)   IMAGE(S)                                 SELECTOR            REPLICAS
frontend       php-redis      gcr.io/google_samples/gb-frontend:v3     name=frontend       3
redis-master   master         redis                                    name=redis-master   1
redis-slave    worker         gcr.io/google_samples/gb-redisslave:v1   name=redis-slave    2
capstone@kube01:~/kubernetes-1.1.7/examples/guestbook$ sudo kubectl get pods
NAME                 READY     STATUS    RESTARTS   AGE
frontend-2blqb       1/1       Running   0          1m
frontend-91v24       1/1       Running   0          1m
frontend-rt4aw       1/1       Running   0          1m
redis-master-k8by5   1/1       Running   0          33m
redis-slave-806w4    1/1       Running   0          11m
redis-slave-bdd04    1/1       Running   0          11m
```

6)  Creating the frontend service

First needed to edit the frontend-service.yaml to use the Nodeport spec to allow external access

```
apiVersion: v1
kind: Service
metadata:
  name: frontend
  labels:
    name: frontend
spec:
  # if your cluster supports it, uncomment the following to automatically create
  # an external load-balanced IP for the frontend service.
  # type: LoadBalancer
  ports:
    # the port that this service should serve on
    - port: 80
    #THE SECTION I ADDED TO ALLOW EXTERNAL ACCESS
    type:NodePort
  selector:
    name: frontend
```

`capstone@kube01:~/kubernetes-1.1.7/examples/guestbook$ sudo kubectl create -f ./frontend-service.yaml`
You have exposed your service on an external port on all nodes in your
cluster.  If you want to expose this service to the external internet, you may
need to set up firewall rules for the service port(s) (tcp:31842) to serve traffic.

See http://releases.k8s.io/HEAD/docs/user-guide/services-firewalls.md for more details.
services/frontend

Determine the Nodeport to access externally

```
capstone@kube01:~/kubernetes-1.1.7/examples/guestbook$ sudo kubectl get svc frontend -o yaml
apiVersion: v1
kind: Service
metadata:
  creationTimestamp: 2016-02-10T14:33:23Z
  labels:
    name: frontend
  name: frontend
  namespace: default
  resourceVersion: "89516"
  selfLink: /api/v1/namespaces/default/services/frontend
  uid: 3d7d23ae-d003-11e5-af23-066e21bab126
spec:
  clusterIP: 192.168.3.189
  ports:
  - nodePort: 31842
    port: 80
    protocol: TCP
    targetPort: 80
  selector:
    name: frontend
  sessionAffinity: None
  type: NodePort
status:
  loadBalancer: {}
capstone@kube01:~/kubernetes-1.1.7/examples/guestbook$ sudo kubectl get servicesNAME           LABELS                                    SELECTOR            IP(S)           PORT(S)
frontend       name=frontend                             name=frontend       192.168.3.138   80/TCP
kubernetes     component=apiserver,provider=kubernetes   <none>              192.168.3.1     443/TCP
redis-master   name=redis-master                         name=redis-master   192.168.3.12    6379/TCP
redis-slave    name=redis-slave                          name=redis-slave    192.168.3.98    6379/TCP
```

7)  Then instructions to kill stuff after we are done

If you are in a live kubernetes cluster, you can just kill the pods by stopping the replication controllers and deleting the services. Using labels to select the resources to stop or delete is an easy way to do this in one command.

```
kubectl stop rc -l "name in (redis-master, redis-slave, frontend)"
kubectl delete service -l "name in (redis-master, redis-slave, frontend)"
```

To completely tear down a Kubernetes cluster, if you ran this from source, you can use:

`$ <kubernetes>/cluster/kube-down.sh`

**Comment by cmische** [comment]
We are having issues exposing services externally.

After spending more time on this than I care to mention, here's what's happening; the nodeport setting does correctly expose externally, but kubernetes is setting up some nat forwarding to properly send the traffic to the container. It's in the pre-routing table, so it changes the destination port for the traffic before it goes into the INPUT chain for our firewall. Then it gets blocked accordingly.

I found this graphic which helped a bit. You can ignore mangle, we don't use it.
![](http://www.ironflake.org/wp-content/uploads/2011/02/Linux_IptablesSchema.png)

I didn't know there was a NAT iptable before yesterday, you can view them like so;
filter table (default): `iptables -L -n`
nat table: `iptables -t nat -L -n`

These are the rules that are doing this

```
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         
KUBE-PORTALS-CONTAINER  all  --  0.0.0.0/0            0.0.0.0/0            /* handle ClusterIPs; NOTE: this must be before the NodePort rules */
KUBE-NODEPORT-CONTAINER  all  --  0.0.0.0/0            0.0.0.0/0            ADDRTYPE match dst-type LOCAL /* handle service NodePorts; NOTE: this must be the last rule in the chain */
DOCKER     all  --  0.0.0.0/0            0.0.0.0/0            ADDRTYPE match dst-type LOCAL
```

```
Chain KUBE-PORTALS-CONTAINER (1 references)
target     prot opt source               destination         
REDIRECT   tcp  --  0.0.0.0/0            192.168.3.12         /* default/redis-master: */ tcp dpt:6379 redir ports 50234
REDIRECT   tcp  --  0.0.0.0/0            192.168.3.98         /* default/redis-slave: */ tcp dpt:6379 redir ports 43971
REDIRECT   udp  --  0.0.0.0/0            192.168.3.10         /* kube-system/kube-dns:dns */ udp dpt:53 redir ports 50535
REDIRECT   tcp  --  0.0.0.0/0            192.168.3.10         /* kube-system/kube-dns:dns-tcp */ tcp dpt:53 redir ports 40767
REDIRECT   tcp  --  0.0.0.0/0            192.168.3.27         /* kube-system/kube-ui: */ tcp dpt:80 redir ports 37269
REDIRECT   tcp  --  0.0.0.0/0            192.168.3.1          /* default/kubernetes: */ tcp dpt:443 redir ports 40566
REDIRECT   tcp  --  0.0.0.0/0            192.168.3.81         /* default/frontend: */ tcp dpt:80 redir ports 39755
```

```
Chain KUBE-NODEPORT-CONTAINER (1 references)
target     prot opt source               destination         
REDIRECT   tcp  --  0.0.0.0/0            0.0.0.0/0            /* default/frontend: */ tcp dpt:30275 redir ports 39755
```

And then port 39755 is one that is not allowed in our firewall. These rules for each service are dynamically created and deleted when corresponding service is created/deleted.

Here's the traffic coming in over the eth1 interface

```
root@kube01:~/kubernetes-1.1.7/examples/guestbook# tcpdump -i eth1 | grep 30000  
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on eth1, link-type EN10MB (Ethernet), capture size 65535 bytes
16:45:39.457935 IP 69-92-45-122.cpe.cableone.net.53849 > kube01.mattbuilt.com.30000: Flags [S], seq 1764203870, win 29200, options [mss 1460,sackOK,TS val 17157248 ecr 0,nop,wscale 7], length 0
16:46:02.859888 IP 69-92-45-122.cpe.cableone.net.53854 > kube01.mattbuilt.com.30000: Flags [S], seq 615567225, win 29200, options [mss 1460,sackOK,TS val 17163099 ecr 0,nop,wscale 7], length 0
16:46:03.857962 IP 69-92-45-122.cpe.cableone.net.53854 > kube01.mattbuilt.com.30000: Flags [S], seq 615567225, win 29200, options [mss 1460,sackOK,TS val 17163349 ecr 0,nop,wscale 7], length 0
16:46:05.874267 IP 69-92-45-122.cpe.cableone.net.53854 > kube01.mattbuilt.com.30000: Flags [S], seq 615567225, win 29200, options [mss 1460,sackOK,TS val 17163850 ecr 0,nop,wscale 7], length 0
```

Here you can see it being blocked repeatedly after it's destination port is changed

```
Feb 14 16:42:23 kube01 kernel: [81196.929029] [UFW BLOCK] IN=eth1 OUT= MAC=06:1f:6e:4c:26:5a:e4:c7:22:63:21:c1:08:00 SRC=69.92.45.122 DST=169.55.21.75 LEN=60 TOS=0x00 PREC=0x00 TTL=46 ID=43887 DF PROTO=TCP SPT=53829 DPT=53339 WINDOW=29200 RES=0x00 SYN URGP=0 
Feb 14 16:42:23 kube01 kernel: [81197.137127] [UFW BLOCK] IN=eth1 OUT= MAC=06:1f:6e:4c:26:5a:e4:c7:22:63:74:41:08:00 SRC=69.92.45.122 DST=169.55.21.75 LEN=60 TOS=0x00 PREC=0x00 TTL=46 ID=52176 DF PROTO=TCP SPT=53827 DPT=53339 WINDOW=29200 RES=0x00 SYN URGP=0 
Feb 14 16:42:23 kube01 kernel: [81197.189101] [UFW BLOCK] IN=eth1 OUT= MAC=06:1f:6e:4c:26:5a:e4:c7:22:63:21:c1:08:00 SRC=69.92.45.122 DST=169.55.21.75 LEN=60 TOS=0x00 PREC=0x00 TTL=46 ID=30615 DF PROTO=TCP SPT=53830 DPT=53339 WINDOW=29200 RES=0x00 SYN URGP=0 
Feb 14 16:42:24 kube01 kernel: [81197.385378] [UFW BLOCK] IN=eth1 OUT= MAC=06:1f:6e:4c:26:5a:e4:c7:22:63:74:41:08:00 SRC=69.92.45.122 DST=169.55.21.75 LEN=60 TOS=0x00 PREC=0x00 TTL=46 ID=60716 DF PROTO=TCP SPT=53828 DPT=53339 WINDOW=29200 RES=0x00 SYN URGP=0 
Feb 14 16:42:24 kube01 kernel: [81198.185372] [UFW BLOCK] IN=eth1 OUT= MAC=06:1f:6e:4c:26:5a:e4:c7:22:63:21:c1:08:00 SRC=69.92.45.122 DST=169.55.21.75 LEN=60 TOS=0x00 PREC=0x00 TTL=46 ID=30616 DF PROTO=TCP SPT=53830 DPT=53339 WINDOW=29200 RES=0x00 SYN URGP=0 
```

Trying to figure out a good solution for this...

Some interesting articles I found in the time looking at this:

https://docs.docker.com/v1.5/articles/networking/
Semi-explains how docker creates containers on ports 49153–65535 and then exposes them.

http://blog.viktorpetersson.com/post/101707677489/the-dangers-of-ufw-docker
Explains that using ufw with docker was at one point pretty dangerous; seems to be fixed? I tried rebooting everything and rebuilding the cluster with "--iptables=false" in /etc/default/docker as it was mentioned in this article and I thought that might stop the iptables port forwarding, but it seems that is implemented by kubernetes. (Rebooting the master without running kube-down.sh required me to rebuild the cluster as the minions were permenantly stuck in a NotReady state) I'll have to put this setting back, default settings are almost always better when trying to debug things.

https://github.com/kubernetes/kubernetes/blob/master/docs/design/networking.md
http://www.systutorials.com/816/port-forwarding-using-iptables/
https://fralef.me/docker-and-iptables.html

I learned a lot at least... still working on fixing this.

A good way to test if a service should be accessible is to first try it on localhost with

`curl localhost:<port>`

If that works, you should be able to access externally via the nodeport by disabling the firewall temporarily.

This is kinda some voodoo magic level stuff :(

**Comment by cmische** [comment]
I think I might have finally found an elegant solution to this. I have a hunch that this wouldn't be an issue at all if we were using a google cloud load balancer, but as of now, that feature is just limited to deployments on the google cloud. Hence we had to use nodeport, and that translation from the nodeport to the actual container port is what was screwing us up.

After reading forever about load balancers I found a stackoverflow post in the corner of the internet

http://stackoverflow.com/a/34622178

That talks about using HostPort, which worked for our guestbook service with ufw

Did some googling on HostPort and found this
https://github.com/kubernetes/kubernetes/issues/3267

I had tried using publicips, but maybe I just set it up wrong? Trying again.

Seems like publicips are depreciated in favor of NodePort. I'm pretty sure NodePort is the way we're supposed to be doing this. The reason I think HostPort worked, is I set HostPort to port 80, and after looking at netstat on the host, there was a kubeproxy process binded to that port. When kube proxy gets that traffic, from my reading, it might route that traffic locally to the service port and basically act as a proxy. I think that's also why we can access the kube-api server externally, it's not running containerized.

Maybe I can just do nodeport and forward all traffic from eth1 into kubeproxy? Don't think so.

I couldn't come up with any good solution and I ended up band aiding it after trying to figure it out for about 10 hours. It seems like nodeport is supposed to just work, I'm not sure why I can't find anyone else complaining about having to open a new ephemeral randomized port in their firewall each time a  service starts. Kubernetes will probably implement load balancers and make this easier in the future and it's not central to our research as long as it works so I'm gonna roll with this band-aid for now.

I just opened up all the ephermeral ports possible that kubernetes might use. 

```
root@kube01:~/kubernetes-1.1.7/examples/guestbook# ufw status verbose
Status: active
Logging: on (high)
Default: deny (incoming), allow (outgoing), deny (routed)
New profiles: skip

To                         Action      From
--                         ------      ----
22                         ALLOW IN    Anywhere
Anywhere                   ALLOW IN    127.0.0.1
Anywhere                   ALLOW IN    169.55.21.75
Anywhere                   ALLOW IN    169.44.62.101
Anywhere                   ALLOW IN    169.44.62.99
80                         ALLOW IN    Anywhere
8080                       ALLOW IN    Anywhere
6443                       ALLOW IN    Anywhere
30000                      ALLOW IN    Anywhere
37000:54000/tcp            ALLOW IN    Anywhere
37000:54000/udp            ALLOW IN    Anywhere
22 (v6)                    ALLOW IN    Anywhere (v6)
37000:54000/udp (v6)       ALLOW IN    Anywhere (v6)

root@kube01:~/kubernetes-1.1.7/examples/guestbook#  

```

I couldn't even find any documentation on what ephermeral ports it uses. It seems to extend past the bounds of docker so that's kind of weird. 

Docker bounds are listed here as well as other places: https://github.com/docker/docker/issues/10220
49153 - 65535

Band-aided for now, we may come back to this and will definitely ask our mentors if they have any advice.

I'll close this issue since it's working and open up a separate one as an enhancement.
